File: fponline/webrtc-signaling.js
// WebRTC Signaling Server Integration
// This handles the signaling needed for WebRTC peer connections

class WebRTCSignalingManager {
  constructor(gameServer) {
    this.gameServer = gameServer;
    this.peerConnections = new Map(); // Track which peers are connected to which
    this.pendingOffers = new Map(); // Track pending offers
  }
  
  handleWebRTCMessage(ws, message) {
    const { type, fromPeer, toPeer, signalType, signalData, clientId } = message;
    
    switch (type) {
      case 'request_webrtc_peers':
        this.handlePeerListRequest(ws, clientId);
        break;
        
      case 'webrtc_signaling':
        this.handleSignaling(ws, fromPeer, toPeer, signalType, signalData);
        break;
        
      default:
        console.log('Unknown WebRTC message type:', type);
    }
  }
  
  handlePeerListRequest(ws, clientId) {
    // Get list of all connected clients except the requester
    const allClients = Array.from(this.gameServer.clients.keys())
      .filter(id => id !== clientId);
    
    console.log(`Sending peer list to ${clientId}:`, allClients);
    
    // Send peer list to the requesting client
    this.sendToClient(ws, {
      type: 'webrtc_signaling',
      fromPeer: 'server',
      toPeer: clientId,
      signalType: 'peer_list',
      signalData: allClients
    });
  }
  
  handleSignaling(ws, fromPeer, toPeer, signalType, signalData) {
    console.log(`WebRTC signaling: ${fromPeer} -> ${toPeer} (${signalType})`);
    
    // Find the target peer's WebSocket connection
    const targetWs = this.findClientWebSocket(toPeer);
    
    if (!targetWs) {
      console.log(`Target peer ${toPeer} not found for signaling`);
      return;
    }
    
    // Forward the signaling message to the target peer
    this.sendToClient(targetWs, {
      type: 'webrtc_signaling',
      fromPeer: fromPeer,
      toPeer: toPeer,
      signalType: signalType,
      signalData: signalData
    });
    
    // Track peer connections
    if (signalType === 'offer') {
      this.pendingOffers.set(`${fromPeer}-${toPeer}`, {
        from: fromPeer,
        to: toPeer,
        timestamp: Date.now()
      });
    } else if (signalType === 'answer') {
      const offerKey = `${toPeer}-${fromPeer}`;
      if (this.pendingOffers.has(offerKey)) {
        this.pendingOffers.delete(offerKey);
        
        // Track successful connection
        this.trackPeerConnection(fromPeer, toPeer);
      }
    }
  }
  
  trackPeerConnection(peer1, peer2) {
    const connectionKey = `${peer1}-${peer2}`;
    this.peerConnections.set(connectionKey, {
      peer1,
      peer2,
      established: Date.now()
    });
    
    console.log(`WebRTC connection established: ${peer1} <-> ${peer2}`);
  }
  
  findClientWebSocket(clientId) {
    // This needs to be adapted to your existing server structure
    // Assuming you have a way to map client IDs to WebSocket connections
    return this.gameServer.clients.get(clientId);
  }
  
  sendToClient(ws, message) {
    if (ws && ws.readyState === 1) { // WebSocket.OPEN
      ws.send(JSON.stringify(message));
    }
  }
  
  handleClientDisconnect(clientId) {
    console.log(`Cleaning up WebRTC connections for ${clientId}`);
    
    // Remove pending offers
    const keysToDelete = [];
    for (const [key, offer] of this.pendingOffers) {
      if (offer.from === clientId || offer.to === clientId) {
        keysToDelete.push(key);
      }
    }
    keysToDelete.forEach(key => this.pendingOffers.delete(key));
    
    // Remove peer connections
    for (const [key, connection] of this.peerConnections) {
      if (connection.peer1 === clientId || connection.peer2 === clientId) {
        this.peerConnections.delete(key);
      }
    }
    
    // Notify other clients that this peer is gone
    this.notifyPeerDisconnection(clientId);
  }
  
  notifyPeerDisconnection(disconnectedPeer) {
    const message = {
      type: 'webrtc_signaling',
      fromPeer: 'server',
      toPeer: 'all',
      signalType: 'peer_disconnected',
      signalData: { peerId: disconnectedPeer }
    };
    
    // Send to all connected clients
    for (const [clientId, ws] of this.gameServer.clients) {
      if (clientId !== disconnectedPeer) {
        this.sendToClient(ws, message);
      }
    }
  }
  
  getStats() {
    return {
      peerConnections: this.peerConnections.size,
      pendingOffers: this.pendingOffers.size,
      connections: Array.from(this.peerConnections.values())
    };
  }
  
  // Clean up old pending offers (call this periodically)
  cleanupOldOffers() {
    const now = Date.now();
    const timeout = 30000; // 30 seconds
    
    const keysToDelete = [];
    for (const [key, offer] of this.pendingOffers) {
      if (now - offer.timestamp > timeout) {
        keysToDelete.push(key);
      }
    }
    
    keysToDelete.forEach(key => {
      console.log(`Cleaning up old offer: ${key}`);
      this.pendingOffers.delete(key);
    });
  }
}

// Example integration with your existing game server
class IntegratedGameServer {
  constructor() {
    this.clients = new Map();
    this.webrtcManager = new WebRTCSignalingManager(this);
    
    // Clean up old WebRTC offers every 30 seconds
    setInterval(() => {
      this.webrtcManager.cleanupOldOffers();
    }, 30000);
  }
  
  handleClientMessage(ws, clientId, message) {
    // Check if this is a WebRTC signaling message
    if (message.type === 'request_webrtc_peers' || message.type === 'webrtc_signaling') {
      this.webrtcManager.handleWebRTCMessage(ws, message);
      return;
    }
    
    // Handle regular game messages
    this.handleGameMessage(ws, clientId, message);
  }
  
  handleGameMessage(ws, clientId, message) {
    // Your existing game logic here
    console.log(`Game message from ${clientId}:`, message.type);
    
    // Example: Handle input messages
    if (message.type === 'input') {
      // Process input and broadcast to other clients
      this.broadcastToOthers(clientId, {
        type: 'player_update',
        playerId: clientId,
        ...message
      });
    }
  }
  
  handleClientDisconnect(clientId) {
    console.log(`Client ${clientId} disconnected`);
    
    // Clean up WebRTC connections
    this.webrtcManager.handleClientDisconnect(clientId);
    
    // Your existing disconnect logic
    this.clients.delete(clientId);
    
    // Notify other clients
    this.broadcastToAll({
      type: 'player_left',
      playerId: clientId
    });
  }
  
  broadcastToOthers(senderId, message) {
    for (const [clientId, ws] of this.clients) {
      if (clientId !== senderId) {
        this.sendToClient(ws, message);
      }
    }
  }
  
  broadcastToAll(message) {
    for (const [clientId, ws] of this.clients) {
      this.sendToClient(ws, message);
    }
  }
  
  sendToClient(ws, message) {
    if (ws && ws.readyState === 1) {
      ws.send(JSON.stringify(message));
    }
  }
  
  getServerStats() {
    return {
      connectedClients: this.clients.size,
      webrtc: this.webrtcManager.getStats()
    };
  }
}

// Example WebSocket server setup
function createWebRTCEnabledServer(port = 8080) {
  const WebSocket = require('ws');
  const gameServer = new IntegratedGameServer();
  
  const wss = new WebSocket.Server({ port });
  
  wss.on('connection', (ws) => {
    const clientId = generateClientId();
    console.log(`Client ${clientId} connected`);
    
    gameServer.clients.set(clientId, ws);
    
    // Send initial connection message
    ws.send(JSON.stringify({
      type: 'init',
      your_id: clientId,
      server_time: Date.now()
    }));
    
    ws.on('message', (data) => {
      try {
        const message = JSON.parse(data);
        gameServer.handleClientMessage(ws, clientId, message);
      } catch (e) {
        console.error('Invalid message format:', e);
      }
    });
    
    ws.on('close', () => {
      gameServer.handleClientDisconnect(clientId);
    });
    
    ws.on('error', (error) => {
      console.error(`WebSocket error for client ${clientId}:`, error);
      gameServer.handleClientDisconnect(clientId);
    });
  });
  
  console.log(`WebRTC-enabled game server listening on port ${port}`);
  return { wss, gameServer };
}

function generateClientId() {
  return Math.random().toString(36).substr(2, 9);
}

module.exports = {
  WebRTCSignalingManager,
  IntegratedGameServer,
  createWebRTCEnabledServer
};

// Usage example:
// const { createWebRTCEnabledServer } = require('./webrtc-signaling');
// const { wss, gameServer } = createWebRTCEnabledServer(8080); 

File: fponline/webrtc_signaling.rs
// WebRTC Signaling Module for Rust Server
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::collections::HashMap;
use tokio::sync::Mutex;
use std::sync::Arc;
use warp::ws::Message;
use futures::SinkExt;
use futures::stream::SplitSink;
use warp::ws::WebSocket;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WebRTCSignalingMessage {
    #[serde(rename = "type")]
    pub message_type: String,
    #[serde(rename = "fromPeer")]
    pub from_peer: Option<u32>,
    #[serde(rename = "toPeer")]
    pub to_peer: Option<u32>,
    #[serde(rename = "signalType")]
    pub signal_type: Option<String>,
    #[serde(rename = "signalData")]
    pub signal_data: Option<Value>,
    #[serde(rename = "clientId")]
    pub client_id: Option<u32>,
}

#[derive(Debug, Clone)]
pub struct PeerConnection {
    pub peer1: u32,
    pub peer2: u32,
    pub established: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone)]
pub struct PendingOffer {
    pub from: u32,
    pub to: u32,
    pub timestamp: chrono::DateTime<chrono::Utc>,
}

pub struct WebRTCSignalingManager {
    pub peer_connections: Arc<Mutex<HashMap<String, PeerConnection>>>,
    pub pending_offers: Arc<Mutex<HashMap<String, PendingOffer>>>,
}

impl WebRTCSignalingManager {
    pub fn new() -> Self {
        Self {
            peer_connections: Arc::new(Mutex::new(HashMap::new())),
            pending_offers: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    pub async fn handle_webrtc_message(
        &self,
        message: &WebRTCSignalingMessage,
        player_id: u32,
        clients: &HashMap<u32, Arc<Mutex<SplitSink<WebSocket, Message>>>>,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        match message.message_type.as_str() {
            "request_webrtc_peers" => {
                self.handle_peer_list_request(player_id, clients).await?;
            }
            "webrtc_signaling" => {
                if let (Some(from_peer), Some(to_peer), Some(signal_type), Some(signal_data)) = (
                    message.from_peer,
                    message.to_peer,
                    &message.signal_type,
                    &message.signal_data,
                ) {
                    self.handle_signaling(from_peer, to_peer, signal_type, signal_data, clients)
                        .await?;
                }
            }
            _ => {
                println!("Unknown WebRTC message type: {}", message.message_type);
            }
        }
        Ok(())
    }

    async fn handle_peer_list_request(
        &self,
        requesting_player: u32,
        clients: &HashMap<u32, Arc<Mutex<SplitSink<WebSocket, Message>>>>,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        // Get list of all connected clients except the requester
        let peer_list: Vec<u32> = clients
            .keys()
            .filter(|&&id| id != requesting_player)
            .copied()
            .collect();

        println!("Sending peer list to {}: {:?}", requesting_player, peer_list);

        // Send peer list to the requesting client
        if let Some(client_ws) = clients.get(&requesting_player) {
            let response = json!({
                "type": "webrtc_signaling",
                "fromPeer": "server",
                "toPeer": requesting_player,
                "signalType": "peer_list",
                "signalData": peer_list
            });

            let mut ws = client_ws.lock().await;
            ws.send(Message::text(response.to_string())).await?;
        }

        Ok(())
    }

    async fn handle_signaling(
        &self,
        from_peer: u32,
        to_peer: u32,
        signal_type: &str,
        signal_data: &Value,
        clients: &HashMap<u32, Arc<Mutex<SplitSink<WebSocket, Message>>>>,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        println!(
            "WebRTC signaling: {} -> {} ({})",
            from_peer, to_peer, signal_type
        );

        // Find the target peer's WebSocket connection
        if let Some(target_ws) = clients.get(&to_peer) {
            // Forward the signaling message to the target peer
            let forward_message = json!({
                "type": "webrtc_signaling",
                "fromPeer": from_peer,
                "toPeer": to_peer,
                "signalType": signal_type,
                "signalData": signal_data
            });

            let mut ws = target_ws.lock().await;
            ws.send(Message::text(forward_message.to_string())).await?;

            // Track peer connections
            match signal_type {
                "offer" => {
                    let offer_key = format!("{}-{}", from_peer, to_peer);
                    let pending_offer = PendingOffer {
                        from: from_peer,
                        to: to_peer,
                        timestamp: chrono::Utc::now(),
                    };
                    self.pending_offers
                        .lock()
                        .await
                        .insert(offer_key, pending_offer);
                }
                "answer" => {
                    let offer_key = format!("{}-{}", to_peer, from_peer);
                    if self.pending_offers.lock().await.remove(&offer_key).is_some() {
                        // Track successful connection
                        self.track_peer_connection(from_peer, to_peer).await;
                    }
                }
                _ => {}
            }
        } else {
            println!("Target peer {} not found for signaling", to_peer);
        }

        Ok(())
    }

    async fn track_peer_connection(&self, peer1: u32, peer2: u32) {
        let connection_key = format!("{}-{}", peer1, peer2);
        let connection = PeerConnection {
            peer1,
            peer2,
            established: chrono::Utc::now(),
        };

        self.peer_connections
            .lock()
            .await
            .insert(connection_key, connection);

        println!("WebRTC connection established: {} <-> {}", peer1, peer2);
    }

    pub async fn handle_client_disconnect(&self, client_id: u32, clients: &HashMap<u32, Arc<Mutex<SplitSink<WebSocket, Message>>>>) {
        println!("Cleaning up WebRTC connections for {}", client_id);

        // Remove pending offers
        let mut pending_offers = self.pending_offers.lock().await;
        pending_offers.retain(|_, offer| offer.from != client_id && offer.to != client_id);

        // Remove peer connections
        let mut peer_connections = self.peer_connections.lock().await;
        peer_connections.retain(|_, connection| {
            connection.peer1 != client_id && connection.peer2 != client_id
        });

        // Notify other clients that this peer is gone
        self.notify_peer_disconnection(client_id, clients).await;
    }

    async fn notify_peer_disconnection(&self, disconnected_peer: u32, clients: &HashMap<u32, Arc<Mutex<SplitSink<WebSocket, Message>>>>) {
        let message = json!({
            "type": "webrtc_signaling",
            "fromPeer": "server",
            "toPeer": "all",
            "signalType": "peer_disconnected",
            "signalData": { "peerId": disconnected_peer }
        });

        // Send to all connected clients
        for (&client_id, client_ws) in clients {
            if client_id != disconnected_peer {
                if let Ok(mut ws) = client_ws.try_lock() {
                    let _ = ws.send(Message::text(message.to_string())).await;
                }
            }
        }
    }

    pub async fn cleanup_old_offers(&self) {
        let now = chrono::Utc::now();
        let timeout = chrono::Duration::seconds(30);

        let mut pending_offers = self.pending_offers.lock().await;
        let initial_count = pending_offers.len();
        
        pending_offers.retain(|key, offer| {
            let should_keep = now.signed_duration_since(offer.timestamp) <= timeout;
            if !should_keep {
                println!("Cleaning up old offer: {}", key);
            }
            should_keep
        });

        let removed_count = initial_count - pending_offers.len();
        if removed_count > 0 {
            println!("Cleaned up {} old WebRTC offers", removed_count);
        }
    }

    pub async fn get_stats(&self) -> (usize, usize) {
        let peer_connections_count = self.peer_connections.lock().await.len();
        let pending_offers_count = self.pending_offers.lock().await.len();
        (peer_connections_count, pending_offers_count)
    }
}

// Helper function to check if a message is a WebRTC signaling message
pub fn is_webrtc_message(text: &str) -> bool {
    text.contains("\"type\":\"request_webrtc_peers\"") 
        || text.contains("\"type\":\"webrtc_signaling\"")
        || text.contains("\"type\": \"request_webrtc_peers\"")
        || text.contains("\"type\": \"webrtc_signaling\"")
}

// Helper function to parse WebRTC signaling message
pub fn parse_webrtc_message(text: &str) -> Result<WebRTCSignalingMessage, serde_json::Error> {
    serde_json::from_str(text)
} 